test/systemtest/utils.py

```diff
@@ -0,0 +1,428 @@
+        self.startup_done_expect = startup_done_expect
+        self.startup_timeout = startup_timeout
+        self.proc = None
+        self.logger = logging.getLogger(__name__)
+
+        self._f_stdout = None
+        self._f_stderr = None
+        self._f_stdout_r = None
+        self._f_stderr_r = None
+
+        self.default_filter_func = default_filter_func
+
+    def __del__(self):
+        try:
+            self.proc.kill()
+            self._f_stdout.close()
+            self._f_stderr.close()
+            self._f_stdout_r.close()
+            self._f_stderr_r.close()
+        except:
+            pass
+
+    def run(self):
+        """Start process with command line configured in constructor."""
+        cmd_name = os.path.basename(self.cmd[0])
+
+        # Enforce line-buffered STDOUT even when sending STDOUT/STDERR to file.
+        # If applications don't fflush() STDOUT manually, STDOUT goes through
+        # a 4kB buffer before we see any output, which prevents searching for
+        # the string indicating a successful startup.
+        # see discussion at http://www.pixelbeat.org/programming/stdio_buffering/
+        cmd = ['stdbuf', '-oL'] + self.cmd
+        self.logger.info("Running command " + ' '.join(cmd))
+
+        logfile_stdout = os.path.join(self.logdir,
+                                      "{}.stdout".format(cmd_name))
+        logfile_stderr = os.path.join(self.logdir,
+                                      "{}.stderr".format(cmd_name))
+        self.logger.debug("Capturing STDOUT at " + logfile_stdout)
+        self.logger.debug("Capturing STDERR at " + logfile_stderr)
+
+        self._f_stdout = open(logfile_stdout, 'w')
+        self._f_stderr = open(logfile_stderr, 'w')
+        self.proc = subprocess.Popen(cmd,
+                                     cwd=self.cwd,
+                                     universal_newlines=True,
+                                     bufsize=1,
+                                     stdin=subprocess.PIPE,
+                                     stdout=self._f_stdout,
+                                     stderr=self._f_stderr)
+
+        self._f_stdout_r = open(logfile_stdout, 'r')
+        self._f_stderr_r = open(logfile_stderr, 'r')
+
+        # no startup match pattern given => startup done!
+        if self.startup_done_expect == None:
+            return True
+
+        # check if the string indicating a successful startup appears in the
+        # the program output (STDOUT or STDERR)
+        init_done = self.find_in_output(pattern=self.startup_done_expect,
+                                        timeout=self.startup_timeout)
+
+        if init_done == None:
+            raise subprocess.TimeoutExpired
+
+        self.logger.info("Startup sequence matched, startup done.")
+
+        return True
+
+    def terminate(self):
+        """Terminates process started by run call."""
+        if not self.proc:
+            return
+        self.proc.terminate()
```

Comment:
This might be easier to read as:
```
if self.proc is not None:
    self.proc.terminate()
```
(no early exit, and it's a line shorter!). Similar for `send_ctrl_c` below.